datafusion/physical-plan/src/repartition/mod.rs

```diff
@@ -298,25 +299,15 @@ impl BatchPartitioner {
                         .into_iter()
                         .enumerate()
                         .filter_map(|(partition, indices)| {
-                            let indices: PrimitiveArray<UInt32Type> = indices.into();
                             (!indices.is_empty()).then_some((partition, indices))
                         })
                         .map(move |(partition, indices)| {
                             // Tracking time required for repartitioned batches construction
                             let _timer = partitioner_timer.timer();
+                            let b: Vec<&RecordBatch> = batches.iter().collect();
 
                             // Produce batches based on indices
-                            let columns = take_arrays(batch.columns(), &indices, None)?;
-
-                            let mut options = RecordBatchOptions::new();
-                            options = options.with_row_count(Some(indices.len()));
-                            let batch = RecordBatch::try_new_with_options(
-                                batch.schema(),
-                                columns,
-                                &options,
-                            )
-                            .unwrap();
-
+                            let batch = interleave_record_batch(&b, &indices)?;
```

Comment:
Probably an api like https://github.com/apache/arrow-rs/pull/7325 would be faster (avoiding one level of "trivial" indexing).

Reply:
FYI @ctsk

Reply:
Nice change, and  a much clearer performance win than https://github.com/apache/datafusion/pull/15479. I expect (without testing) that these two PRs interact negatively with one another - Removing coalesce will mean that the data is "more scattered" in memory and probably make interleave work worse - as well as the computation of the left join keys.

Reply:
I think removing coalesce after this change (for all hash repartitions) might be possible, as the output batch size will be roughly equal to input batch size (instead of roughly 1/partitions * batch_size). Unless hash values are somehow skewed (but this is currently also not good anyway).

A future api could use your `take_in` api maybe to only output rows once batch size has been reached.

Reply:
I see I had misunderstood this PR. It makes a lot of sense to do this. As part of prototyping the integration of a take_in in API in datafusion, I made a similar change - move the buffering  *before* sending to their destination thread. I don't remeber see as much speedup when I benchmarked that change independently - I guess using interleave instead of a take/concat (like I did back then) combo makes a significant difference. Awesome!

Reply:
Yeah the speedup comes from avoiding copying the data a second time in `concat` / CoalesceBatches. So when using `take_in` we should be careful to use it once (for a single destination batch) to avoid doing the concat on the small batches